inline-r/src/Foreign/R.hsc

```diff
@@ -155,206 +154,219 @@ module Foreign.R
 -- Environment functions                                                      --
 --------------------------------------------------------------------------------
 
--- | Environment frame.
-{# fun FRAME as envFrame { unsexp `SEXP s R.Env' } -> `SEXP s R.PairList' sexp #}
+envFrame :: (SEXP s 'R.Env) -> IO (SEXP s R.PairList)
+envFrame (unsexp -> s) = sexp <$> [C.exp| SEXP { FRAME( $(SEXP s) ) } |]
 
 -- | Enclosing environment.
-{# fun ENCLOS as envEnclosing { unsexp `SEXP s R.Env' } -> `SEXP s R.Env' sexp #}
+envEnclosing :: SEXP s 'R.Env -> IO (SEXP s 'R.Env)
+envEnclosing (unsexp -> s) = sexp <$> [C.exp| SEXP { ENCLOS( $(SEXP s) ) } |]
 
 -- | Hash table associated with the environment, used for faster name lookups.
-{# fun HASHTAB as envHashtab { unsexp `SEXP s R.Env' } -> `SEXP s R.Vector' sexp #}
+envHashtab :: SEXP s 'R.Env -> IO (SEXP s 'R.Vector)
+envHashtab (unsexp -> s) = sexp <$> [C.exp| SEXP { HASHTAB( $(SEXP s) ) } |]
 
 --------------------------------------------------------------------------------
 -- Closure functions                                                          --
 --------------------------------------------------------------------------------
 
 -- | Closure formals (aka the actual arguments).
-{# fun FORMALS as closureFormals { unsexp `SEXP s R.Closure' } -> `SEXP s R.PairList' sexp #}
+closureFormals :: SEXP s 'R.Closure -> IO (SEXP s R.PairList)
+closureFormals (unsexp -> s) = sexp <$> [C.exp| SEXP { FORMALS( $(SEXP s) ) }|]
 
 -- | The code of the closure.
-{# fun BODY as closureBody { unsexp `SEXP s R.Closure' } -> `SomeSEXP s' somesexp #}
+closureBody :: SEXP s 'R.Closure -> IO (SomeSEXP s)
+closureBody (unsexp -> s) = somesexp <$> [C.exp| SEXP { BODY( $(SEXP s) ) } |]
 
 -- | The environment of the closure.
-{# fun CLOENV as closureEnv { unsexp `SEXP s R.Closure' } -> `SEXP s R.Env' sexp #}
+closureEnv :: SEXP s 'R.Closure -> IO (SEXP s 'R.Env)
+closureEnv (unsexp -> s) = sexp <$> [C.exp| SEXP { CLOENV( $(SEXP s) ) }|]
 
 --------------------------------------------------------------------------------
 -- Promise functions                                                          --
 --------------------------------------------------------------------------------
 
 -- | The code of a promise.
-{# fun PRCODE as promiseCode { unsexp `SEXP s R.Promise'} -> `SomeSEXP s' somesexp #}
+promiseCode :: SEXP s 'R.Promise -> IO (SomeSEXP s)
+promiseCode (unsexp -> s) = somesexp <$> [C.exp| SEXP { PRVALUE( $(SEXP s) )}|]
 
 -- | The environment in which to evaluate the promise.
-{# fun PRENV as promiseEnv { unsexp `SEXP s R.Promise'} -> `SEXP s R.Env' sexp #}
+promiseEnv :: SEXP s 'R.Promise -> IO (SomeSEXP s)
+promiseEnv (unsexp -> s) = somesexp <$> [C.exp| SEXP { PRENV( $(SEXP s) )}|]
 
 -- | The value of the promise, if it has already been forced.
-{# fun PRVALUE as promiseValue { unsexp `SEXP s R.Promise'} -> `SomeSEXP s' somesexp #}
+promiseValue :: SEXP s 'R.Promise -> IO (SomeSEXP s)
```

Comment:
Careful with what functions you're binding. Above: `promiseCode -> PRVALUE` and `promiseValue -> PRCODE`. I think you meant the opposite mapping.